src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,144 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node heap;
+    private Node tall;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            Node<T> newNode = new Node<T>(null, value, null);
+            heap = newNode;
+            tall = newNode;
+        } else {
+            Node<T> newNode = new Node<T>(tall, value, null);
+            tall.next = newNode;
+            tall = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index was failed ");
+        }
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, heap);
+            heap.prev = newNode;
+            heap = newNode;
+            size++;
+        } else {
+            Node<T> newNode = new Node<T>(getNode(index - 1), value, getNode(index));
+            getNode(index).prev = newNode;
+            getNode(index - 1).next = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> result = heap;
+        for (int i = 0; i < index; i++) {
+            result = result.next;
+        }
+        return result.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T result = getNode(index).item;
+        getNode(index).item = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size || size == 0) {
+            throw new IndexOutOfBoundsException("Index was failed ");
+        }
+        T result = getNode(index).item;
+        Node<T> indexNode = getNode(index);
+        if (index == 0 && size > 1) {
+            heap = indexNode.next;
+            indexNode.next.prev = null;
+            size--;
+            return result;
+        } else if (index == 0) {
+            indexNode.item = null;
+            size--;
+            heap = null;
+            tall = null;
+            return result;
+        } else if (index == size - 1) {
+            tall = indexNode.prev;
+            indexNode.prev = new Node<>(indexNode.prev.prev, indexNode.prev.item, null);
+            size--;
+            return result;
+        }
+        indexNode.prev.next = indexNode.next;
+        indexNode.next.prev = indexNode.prev;
+        size--;
+        return indexNode.item;
```

Comment:
let's move this part of the code to a separate method, and then we can call this method from remove(index) or remove(object) methods